lib/libutee/tee_api.c

```diff
@@ -16,36 +16,102 @@ static const void *tee_api_instance_data;
 
 /* System API - Internal Client API */
 
-void __utee_from_param(struct utee_params *up, uint32_t param_types,
-			const TEE_Param params[TEE_NUM_PARAMS])
+static TEE_Result copy_param(struct utee_params *up, uint32_t param_types,
+			     const TEE_Param params[TEE_NUM_PARAMS],
+			     void **tmp_buf, size_t *tmp_len,
+			     void *tmp_va[TEE_NUM_PARAMS])
 {
-	size_t n;
+	size_t n = 0;
+	uint8_t *tb = NULL;
+	size_t tbl = 0;
+	size_t tmp_align = sizeof(vaddr_t) * 2;
+	bool is_tmp_mem[TEE_NUM_PARAMS] = { false };
+	void *b = NULL;
+	size_t s = 0;
+	const uint32_t flags = TEE_MEMORY_ACCESS_READ;
+
+	/*
+	 * If a memory parameter points to TA private memory we need to
+	 * allocate a temporary buffer to avoid exposing the memory
+	 * directly to the called TA.
+	 */
+
+	*tmp_buf = NULL;
+	*tmp_len = 0;
+	for (n = 0; n < TEE_NUM_PARAMS; n++) {
+		tmp_va[n] = NULL;
+		switch (TEE_PARAM_TYPE_GET(param_types, n)) {
+		case TEE_PARAM_TYPE_MEMREF_INPUT:
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+			b = params[n].memref.buffer;
+			s = params[n].memref.size;
+			/*
+			 * We're only allocating temporary memory if the
+			 * buffer is completely within TA memory. If it's
+			 * NULL, empty, partiall outside or completely
+			 * outside TA memory there's nothing more we need
+			 * to do here. If there's security/permissions
+			 * problem we'll get an error in the
+			 * invoke_command/open_session below.
+			 */
+			if (b && s &&
+			    !TEE_CheckMemoryAccessRights(flags, b, s)) {
```

Comment:
minor: would prefer `if (b && s && TEE_CheckMemoryAccessRights(flags, b, s) == TEE_SUCCESS)` which is more explicit than `if (b && s && !TEE_CheckMemoryAccessRights(flags, b, s))`.

Reply:
This is a matter of taste. The two expressions are equivalent and both styles are OK with our coding style. The style used here is normally the preferred style so I'd rather keep it as it is.